# Optimised SQL Queries: Flights Database
© Explore Data Science Academy

## Instructions to Students

This challenge is designed to determine how much you have learned so far and will test your knowledge on Optimised SQL queries.

The answers for this challenge should be selected on Athena for each corresponding Multiple Choice Question. The questions are included in this notebook and are numbered according to the Athena Questions, the options to choose from for each question has also been included.

Do not add or remove cells in this notebook. Do not edit or remove the `%%sql` comment as it is required to run each cell.

**_Good Luck!_**

## Honour Code

I YOUR NAME, YOUR SURNAME, confirm - by submitting this document - that the solutions in this notebook are a result of my own work and that I abide by the EDSA honour code (https://drive.google.com/file/d/1QDCjGZJ8-FmJE3bZdIQNwnJyQKPhHZBn/view?usp=sharing).

Non-compliance with the honour code constitutes a material breach of contract.

## The flights database

Optimised queries are vital for increasing database performance, saving resources and reducing network overhead. In this challenge we will be  using actual US flights data to illustrate the use of optimisation techniques with large amounts of data. This database consists of information relating to flights, carriers, airports and planes. 

<img src="https://github.com/Explore-AI/Pictures/blob/master/Flights.jpg?raw=true" width=50%/> 


Below is an Entity Relationship Diagram (ERD) of the flights database:

<img src="https://github.com/Explore-AI/Pictures/blob/master/flights_ER_diagram.png?raw=true" width=40%/>

As can be seen from the ERD, the flights database consists of four tables:
- **flights**: all domestic flights in the USA in 2008
- **carriers**: lookup table for all the carriers
- **planes**: lookup table for all the planes 
- **airports**: lookup table for all the airports

For this challenge you will apply the use of optimised SQL queries to gain insight into the flights database. 

## Loading the database

Before you start producing optimised SQL queries, you need to prepare your SQL environment. You can do this by loading in the magic command `%load_ext sql` as we've seen in previous trains. Next you can go ahead and load in your database. To do this you will need to ensure you have downloaded the `flight.db`sqlite file from Athena and have stored it in a known location.  

In [1]:
%load_ext sql

In [2]:
%%sql 

sqlite:///flights.db

## Exploring the database 

Before we begin this challenge we can go ahead and take a look at the composition of each table in the flights database

In [3]:
%%sql 

PRAGMA table_info(flights);

 * sqlite:///flights.db
Done.


[(0, 'index', 'INTEGER', 0, 'NULL', 0),
 (1, 'Date', 'longtext', 0, None, 0),
 (2, 'DayOfWeek', 'INTEGER', 0, 'NULL', 0),
 (3, 'DepTime', 'double', 0, 'NULL', 0),
 (4, 'CRSDepTime', 'longtext', 0, None, 0),
 (5, 'ArrTime', 'double', 0, 'NULL', 0),
 (6, 'CRSArrTime', 'longtext', 0, None, 0),
 (7, 'UniqueCarrier', 'longtext', 0, None, 0),
 (8, 'FlightNum', 'longtext', 0, None, 0),
 (9, 'TailNum', 'longtext', 0, None, 0),
 (10, 'ActualElapsedTime', 'double', 0, 'NULL', 0),
 (11, 'CRSElapsedTime', 'longtext', 0, None, 0),
 (12, 'AirTime', 'longtext', 0, None, 0),
 (13, 'ArrDelay', 'double', 0, 'NULL', 0),
 (14, 'DepDelay', 'double', 0, 'NULL', 0),
 (15, 'Origin', 'longtext', 0, None, 0),
 (16, 'Dest', 'longtext', 0, None, 0),
 (17, 'Distance', 'INTEGER', 0, 'NULL', 0),
 (18, 'TaxiIn', 'longtext', 0, None, 0),
 (19, 'TaxiOut', 'longtext', 0, None, 0),
 (20, 'Cancelled', 'longtext', 0, None, 0),
 (21, 'CancellationCode', 'longtext', 0, None, 0),
 (22, 'Diverted', 'longtext', 0, None, 0),
 (23, 'CarrierDelay', 'longtext', 0, None, 0),
 (24, 'WeatherDelay', 'longtext', 0, None, 0),
 (25, 'NASDelay', 'longtext', 0, None, 0),
 (26, 'SecurityDelay', 'longtext', 0, None, 0),
 (27, 'LateAircraftDelay', 'longtext', 0, None, 0)]

In [4]:
%%sql 

PRAGMA table_info(carriers);

 * sqlite:///flights.db
Done.


[(0, 'Code', 'varchar(10)', 0, 'NULL', 0),
 (1, 'Description', 'varchar(100)', 0, 'NULL', 0)]

In [5]:
%%sql 

PRAGMA table_info(airports);

 * sqlite:///flights.db
Done.


[(0, 'iata', 'varchar(10)', 0, 'NULL', 0),
 (1, 'airport', 'varchar(100)', 0, 'NULL', 0),
 (2, 'city', 'varchar(100)', 0, 'NULL', 0),
 (3, 'state', 'varchar(50)', 0, 'NULL', 0),
 (4, 'country', 'varchar(50)', 0, 'NULL', 0),
 (5, 'lat', 'float(24,0)', 0, 'NULL', 0),
 (6, 'long', 'float(24,0)', 0, 'NULL', 0)]

In [6]:
%%sql 

PRAGMA table_info(planes);

 * sqlite:///flights.db
Done.


[(0, 'tailnum', 'varchar(10)', 0, 'NULL', 0),
 (1, 'type', 'varchar(100)', 0, 'NULL', 0),
 (2, 'manufacturer', 'varchar(100)', 0, 'NULL', 0),
 (3, 'issue_date', 'varchar(100)', 0, 'NULL', 0),
 (4, 'model', 'varchar(50)', 0, 'NULL', 0),
 (5, 'status', 'varchar(100)', 0, 'NULL', 0),
 (6, 'aircraft_type', 'varchar(100)', 0, 'NULL', 0),
 (7, 'engine_type', 'varchar(100)', 0, 'NULL', 0),
 (8, 'year', 'longtext', 0, None, 0)]

## Questions on Optimised SQL Queries 

Now that you have an idea of what information each table contains, you can use the given cell below each question to execute your optimised SQL queries to find the correct output from the options provided for the multiple choice questions on Athena.

**Question 1**

How many different carriers are there in total in the database?

**Options:**
- 1 252
- 1491
- 20
- 37

In [10]:
%%sql

SELECT COUNT(DISTINCT(Description)) FROM carriers AS No_of_carriers;

 * sqlite:///flights.db
Done.


[(1491,)]

**Question 2** 

How long was the longest delay before departure? 

**Options:**
- 1355 min
- 999 min
- 1099 min
- 588 min

In [15]:
%%sql 

SELECT MAX(DepDelay) FROM flights AS longest_delay; 


 * sqlite:///flights.db
Done.


[(1355.0,)]

**Question 3** 

How many flights departed on the 28th of January 2008?

**Options:**
- 540 908
- 20 149
- 19 495
- 18 056

In [28]:
%%sql 

With CTE AS
(SELECT FlightNum, Date, Cancelled FROM Flights)
SELECT COUNT(FlightNum), Date
FROM CTE
WHERE Date = '2008/1/28'
    AND Cancelled = '0';

 * sqlite:///flights.db
Done.


[(19495, '2008/1/28')]

**Question 4**
 
 What is the distance between Midway Airport (MDW) and Houston Airport (HOU)?
 
 **Options:**
 - 611 km
 - 972 km
 - 1 121 km
 - 937 km

In [32]:
%%sql

With CTE AS
(SELECT Origin, Dest, Distance FROM Flights)

SELECT * FROM CTE
WHERE Origin = 'MDW'
    AND Dest = 'HOU'
LIMIT 1;

 * sqlite:///flights.db
Done.


[('MDW', 'HOU', 937)]

**Question 5**

Which day of the week had the highest number of cancelled flights?(Note 1 = cancelled , 0 = not cancelled;)


**Options:**
- Monday(1)
- Wednesday(3) 
- Saturday(6) 
- Thursday (4) 

In [38]:
%%sql 

With CTE AS
(SELECT DayOfWeek, Cancelled FROM Flights)

SELECT COUNT(DayOfWeek), DayOfWeek FROM CTE
WHERE Cancelled = '1'
GROUP BY DayOfWeek;

 * sqlite:///flights.db
Done.


[(2617, 1), (2993, 2), (2645, 3), (3093, 4), (2049, 5), (1984, 6), (1927, 7)]

**Question 6**

How many airports have the word "International" in their name?

**Options:**
- 110
- 124
- 8
- 2

In [41]:
%%sql

SELECT COUNT(airport) FROM airports
WHERE airport LIKE '%international%';

 * sqlite:///flights.db
Done.


[(124,)]

**Question 7**

What is the most produced model for the manufacturer "BOEING"?

**Options:**
- 717-200
- 737-7H4
- 757-222
- 737-3H4

In [55]:
%%sql 

With CTE AS
(SELECT manufacturer, model FROM planes)

SELECT manufacturer, model, COUNT(model) FROM CTE 
WHERE manufacturer = 'BOEING'
GROUP BY model
ORDER BY COUNT(model) DESC;

 * sqlite:///flights.db
Done.


[('BOEING', '737-7H4', 308),
 ('BOEING', '737-3H4', 147),
 ('BOEING', '757-232', 112),
 ('BOEING', '737-824', 100),
 ('BOEING', '717-200', 98),
 ('BOEING', '757-222', 96),
 ('BOEING', '767-332', 77),
 ('BOEING', '737-832', 68),
 ('BOEING', '737-322', 64),
 ('BOEING', '737-524', 56),
 ('BOEING', '777-222', 52),
 ('BOEING', '737-3TO', 50),
 ('BOEING', '757-251', 49),
 ('BOEING', '757-224', 41),
 ('BOEING', '737-724', 36),
 ('BOEING', '767-322', 35),
 ('BOEING', '737-890', 33),
 ('BOEING', '737-7BD', 32),
 ('BOEING', '737-522', 30),
 ('BOEING', '767-323', 27),
 ('BOEING', '757-223', 27),
 ('BOEING', '737-76N', 27),
 ('BOEING', '737-4B7', 27),
 ('BOEING', '737-5H4', 25),
 ('BOEING', '747-422', 23),
 ('BOEING', '737-4Q8', 22),
 ('BOEING', '767-432ER', 21),
 ('BOEING', '737-3G7', 20),
 ('BOEING', '737-790', 19),
 ('BOEING', '757-2B7', 18),
 ('BOEING', '767-424ER', 16),
 ('BOEING', '757-351', 16),
 ('BOEING', '737-490', 16),
 ('BOEING', '737-3B7', 16),
 ('BOEING', '767-223', 15),
 ('BOEING', '757-225', 13),
 ('BOEING', '737-401', 13),
 ('BOEING', '777-224', 12),
 ('BOEING', '737-990', 12),
 ('BOEING', '737-924', 12),
 ('BOEING', '767-224', 10),
 ('BOEING', '757-324', 9),
 ('BOEING', '757-2Q8', 9),
 ('BOEING', '737-3Q8', 9),
 ('BOEING', '737-3A4', 9),
 ('BOEING', '777-232', 8),
 ('BOEING', '767-33A', 8),
 ('BOEING', '757-33N', 8),
 ('BOEING', '737-924ER', 8),
 ('BOEING', '737-301', 7),
 ('BOEING', '767-3P6', 6),
 ('BOEING', '757-2S7', 6),
 ('BOEING', '757-231', 6),
 ('BOEING', '737-3Y0', 6),
 ('BOEING', '767-2B7', 5),
 ('BOEING', '767-201', 5),
 ('BOEING', '757-212', 4),
 ('BOEING', '737-3S3', 4),
 ('BOEING', '737-33A', 4),
 ('BOEING', '737-317', 4),
 ('BOEING', '737-236', 4),
 ('BOEING', '767-3G5', 3),
 ('BOEING', '767-3CB', 3),
 ('BOEING', '757-2G7', 3),
 ('BOEING', '757-23N', 3),
 ('BOEING', '737-3T5', 3),
 ('BOEING', '777-232LR', 2),
 ('BOEING', '737-7Q8', 2),
 ('BOEING', '737-76Q', 2),
 ('BOEING', '737-73A', 2),
 ('BOEING', '737-3L9', 2),
 ('BOEING', '737-3K2', 2),
 ('BOEING', '767-324', 1),
 ('BOEING', '757-26D', 1),
 ('BOEING', '747-451', 1),
 ('BOEING', '747-2B5F', 1),
 ('BOEING', '737-8FH', 1),
 ('BOEING', '737-7AD', 1),
 ('BOEING', '737-705', 1),
 ('BOEING', '737-4S3', 1),
 ('BOEING', '737-2Y5', 1),
 ('BOEING', '737-2X6C', 1),
 ('BOEING', '737-2P6', 1),
 ('BOEING', '737-282C', 1),
 ('BOEING', '737-282', 1),
 ('BOEING', '737-230', 1)]

**Question 8**

What manufacturer had the highest average delay time (DepDelay + ArrDelay)? 

**Options:**
- PAIR MIKE E
- AERONCA
- AVIAT AIRCRAFT INC
- BOEING OF CANADA LTD

In [ ]:
%%sql

SELECT manufacturer, DepTime, ArrDelay
FROM flights
INNER JOIN planes
ON flights.tailnum = planes.tailnum

 * sqlite:///flights.db
Done.


**Question 9**

How many planes landed at Los Angeles International Airport? 

**Options:**
- 215 608
- 18 514
- 41 258
- 39 422

In [71]:
%%sql 

SELECT airport, iata, COUNT(Dest)
FROM airports
INNER JOIN flights
ON airports.iata = flights.Dest
WHERE airport = 'Los Angeles International'
    AND Cancelled = '0';


 * sqlite:///flights.db
Done.


[('Los Angeles International', 'LAX', 18514)]

**Question 10**

Which domestic carrier had the best on-time performance (OTP), where OTP is defined as the rate of on time flights with a 15min buffer on departure and arrival?

**Options:**
- American Airlines Inc
- Hawaiian Airlines Inc
- Comair Inc
- Aloha Airlines Inc

In [73]:
%%sql 

SELECT Description, DepTime, ArrTime
FROM flights
INNER JOIN carriers
ON flights.UniqueCarrier = carriers.Code
WHERE ArrTime - DepTime = 15

 * sqlite:///flights.db
Done.


[('Southwest Airlines Co.\r', 644.0, 659.0),
 ('Southwest Airlines Co.\r', 706.0, 721.0),
 ('Southwest Airlines Co.\r', 1927.0, 1942.0),
 ('Southwest Airlines Co.\r', 1510.0, 1525.0),
 ('Southwest Airlines Co.\r', 2231.0, 2246.0),
 ('Southwest Airlines Co.\r', 1723.0, 1738.0),
 ('Southwest Airlines Co.\r', 2135.0, 2150.0),
 ('Southwest Airlines Co.\r', 618.0, 633.0),
 ('Southwest Airlines Co.\r', 1536.0, 1551.0),
 ('Southwest Airlines Co.\r', 1729.0, 1744.0),
 ('Southwest Airlines Co.\r', 1541.0, 1556.0),
 ('Southwest Airlines Co.\r', 1137.0, 1152.0),
 ('Southwest Airlines Co.\r', 2234.0, 2249.0),
 ('Southwest Airlines Co.\r', 807.0, 822.0),
 ('Southwest Airlines Co.\r', 707.0, 722.0),
 ('Southwest Airlines Co.\r', 2212.0, 2227.0),
 ('Southwest Airlines Co.\r', 1238.0, 1253.0),
 ('Southwest Airlines Co.\r', 1721.0, 1736.0),
 ('Southwest Airlines Co.\r', 1811.0, 1826.0),
 ('Southwest Airlines Co.\r', 1743.0, 1758.0),
 ('Southwest Airlines Co.\r', 1021.0, 1036.0),
 ('Southwest Airlines Co.\r', 830.0, 845.0),
 ('Southwest Airlines Co.\r', 743.0, 758.0),
 ('Southwest Airlines Co.\r', 1622.0, 1637.0),
 ('Southwest Airlines Co.\r', 1916.0, 1931.0),
 ('Southwest Airlines Co.\r', 1305.0, 1320.0),
 ('Southwest Airlines Co.\r', 1726.0, 1741.0),
 ('Southwest Airlines Co.\r', 1441.0, 1456.0),
 ('Southwest Airlines Co.\r', 1814.0, 1829.0),
 ('Southwest Airlines Co.\r', 1315.0, 1330.0),
 ('Southwest Airlines Co.\r', 2120.0, 2135.0),
 ('Southwest Airlines Co.\r', 1843.0, 1858.0),
 ('Southwest Airlines Co.\r', 631.0, 646.0),
 ('Southwest Airlines Co.\r', 835.0, 850.0),
 ('Southwest Airlines Co.\r', 2032.0, 2047.0),
 ('Southwest Airlines Co.\r', 706.0, 721.0),
 ('Southwest Airlines Co.\r', 1606.0, 1621.0),
 ('Southwest Airlines Co.\r', 1129.0, 1144.0),
 ('Southwest Airlines Co.\r', 701.0, 716.0),
 ('Southwest Airlines Co.\r', 1428.0, 1443.0),
 ('Southwest Airlines Co.\r', 830.0, 845.0),
 ('Southwest Airlines Co.\r', 1830.0, 1845.0),
 ('Southwest Airlines Co.\r', 1003.0, 1018.0),
 ('Southwest Airlines Co.\r', 2018.0, 2033.0),
 ('Southwest Airlines Co.\r', 902.0, 917.0),
 ('Southwest Airlines Co.\r', 1434.0, 1449.0),
 ('Southwest Airlines Co.\r', 1939.0, 1954.0),
 ('Southwest Airlines Co.\r', 2208.0, 2223.0),
 ('Southwest Airlines Co.\r', 1019.0, 1034.0),
 ('Southwest Airlines Co.\r', 1224.0, 1239.0),
 ('Southwest Airlines Co.\r', 1830.0, 1845.0),
 ('Southwest Airlines Co.\r', 2007.0, 2022.0),
 ('Southwest Airlines Co.\r', 1830.0, 1845.0),
 ('Southwest Airlines Co.\r', 1706.0, 1721.0),
 ('Southwest Airlines Co.\r', 1813.0, 1828.0),
 ('Southwest Airlines Co.\r', 1807.0, 1822.0),
 ('Southwest Airlines Co.\r', 1909.0, 1924.0),
 ('Southwest Airlines Co.\r', 1234.0, 1249.0),
 ('Southwest Airlines Co.\r', 1105.0, 1120.0),
 ('Southwest Airlines Co.\r', 1630.0, 1645.0),
 ('Southwest Airlines Co.\r', 919.0, 934.0),
 ('Southwest Airlines Co.\r', 1616.0, 1631.0),
 ('Southwest Airlines Co.\r', 901.0, 916.0),
 ('Southwest Airlines Co.\r', 2039.0, 2054.0),
 ('Southwest Airlines Co.\r', 1939.0, 1954.0),
 ('Southwest Airlines Co.\r', 1016.0, 1031.0),
 ('Southwest Airlines Co.\r', 1802.0, 1817.0),
 ('Southwest Airlines Co.\r', 1139.0, 1154.0),
 ('Southwest Airlines Co.\r', 903.0, 918.0),
 ('Southwest Airlines Co.\r', 1803.0, 1818.0),
 ('Southwest Airlines Co.\r', 1227.0, 1242.0),
 ('Southwest Airlines Co.\r', 1014.0, 1029.0),
 ('Southwest Airlines Co.\r', 1219.0, 1234.0),
 ('Southwest Airlines Co.\r', 714.0, 729.0),
 ('Southwest Airlines Co.\r', 1817.0, 1832.0),
 ('Southwest Airlines Co.\r', 2120.0, 2135.0),
 ('Southwest Airlines Co.\r', 1119.0, 1134.0),
 ('Southwest Airlines Co.\r', 1220.0, 1235.0),
 ('Southwest Airlines Co.\r', 906.0, 921.0),
 ('Southwest Airlines Co.\r', 2213.0, 2228.0),
 ('Southwest Airlines Co.\r', 1810.0, 1825.0),
 ('Southwest Airlines Co.\r', 1024.0, 1039.0),
 ('Southwest Airlines Co.\r', 1941.0, 1956.0),
 ('Southwest Airlines Co.\r', 1631.0, 1646.0),
 ('Southwest Airlines Co.\r', 842.0, 857.0),
 ('Southwest Airlines Co.\r', 1505.0, 1520.